crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs

```diff
@@ -5,6 +5,17 @@ unsafe fn test_fd_dir_ops(dir_fd: wasi::Fd) {
     let stat = wasi::fd_filestat_get(dir_fd).expect("failed to fdstat");
     assert_eq!(stat.filetype, wasi::FILETYPE_DIRECTORY);
 
+    let fd = (3..)
+        .map_while(|fd| {
+            wasi::fd_prestat_get(fd)
+                .ok()
+                .map(|wasi::Prestat { tag, .. }| (fd, tag))
+        })
+        .find_map(|(fd, tag)| (tag == wasi::PREOPENTYPE_DIR.raw()).then_some(fd))
+        .expect("failed to find preopen directory");
+    let r = wasi::fd_prestat_dir_name(fd, [].as_mut_ptr(), 0);
+    assert_eq!(r, Err(wasi::ERRNO_NAMETOOLONG));
```

Comment:
Can we also add a case where we pass a buffer which is 1 byte longer than required